python/rikai/spark/sql/codegen/mlflow_logger.py

```diff
@@ -0,0 +1,123 @@
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Custom Mlflow model logger to make sure models have the right logging for
+Rikai SQL ML
+"""
+from typing import Any, Optional
+
+import mlflow
+
+from rikai.conf import (
+    CONF_MLFLOW_ARTIFACT_PATH,
+    CONF_MLFLOW_MODEL_FLAVOR,
+    CONF_MLFLOW_OUTPUT_SCHEMA,
+    CONF_MLFLOW_POST_PROCESSING,
+    CONF_MLFLOW_PRE_PROCESSING,
+    CONF_MLFLOW_SPEC_VERSION,
+    CONF_MLFLOW_TRACKING_URI,
+)
+
+KNOWN_FLAVORS = ["pytorch"]
+__all__ = KNOWN_FLAVORS
+
+
+class MlflowLogger:
+    """
+    An alternative model logger for use during training instead of the vanilla
+
+    """
+
+    _CURRENT_MODEL_SPEC_VERSION = "1.0"
+
+    def __init__(self, flavor: str):
+        """
+        Parameters
+        ----------
+        flavor: str
+            The model flavor
+        """
+        self.spec_version = MlflowLogger._CURRENT_MODEL_SPEC_VERSION
+        self.flavor = flavor
+
+    def log_model(
+        self,
+        model: Any,
+        artifact_path: str,
+        schema: str,
+        pre_processing: Optional[str] = None,
+        post_processing: Optional[str] = None,
+        registered_model_name: Optional[str] = None,
+        **kwargs,
+    ):
+        """Convenience function to log the model with tags needed by rikai.
+        This should be called during training when the model is produced.
+
+        Parameters
+        ----------
+        model: Any
+            The model artifact object
+        artifact_path: str
+            The relative (to the run) artifact path
+        schema: str
+            Output schema (pyspark DataType)
+        pre_processing: str, default None
+            Full python module path of the pre-processing transforms
+        post_processing: str, default None
+            Full python module path of the post-processing transforms
+        registered_model_name: str, default None
+            Model name in the mlflow model registry
+        kwargs: dict
+            Passed to `mlflow.<flavor>.log_model`
+
+        Examples
+        --------
+        # import mlflow.pytorch  <= vanilla mlflow usage
+        import rikai.mlflow
+
+        # Log PyTorch model
+        with mlflow.start_run() as run:
+
+            # Training loop
+            # ...
+
+            # Assume `model` is the trained model from the training loop
+            rikai.mlflow.pytorch.log_model(model, "model",
+                      schema=rikai.contrib.torch.transforms.fasterrcnn_resnet50_fpn.OUTPUT_SCHEMA,
+                      pre_processing="rikai.contrib.torch.transforms.fasterrcnn_resnet50_fpn.pre_processing",
```

Comment:
is it possible that we take this parameter as a function instead of a string?

a string could make refactory much harder.

Reply:
so turn the function into a string inside log_model then write it to the server?